test/extended/single_node/resiliency.go

```diff
@@ -0,0 +1,91 @@
+package single_node
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	v1 "github.com/openshift/api/config/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	"github.com/openshift/origin/pkg/monitor"
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/openshift/origin/test/extended/util/disruption"
+	"github.com/openshift/origin/test/extended/util/disruption/controlplane"
+	types "k8s.io/apimachinery/pkg/types"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+)
+
+func forceApiRollback(oc *exutil.CLI) {
+	redeploymentReason := fmt.Sprintf(`{"spec":{"forceRedeploymentReason":"resiliency-test-%v"}}`, time.Now().Format(time.RFC3339))
+
+	_, err := oc.AdminOperatorClient().OperatorV1().KubeAPIServers().Patch(context.Background(), "cluster", types.MergePatchType,
+		[]byte(redeploymentReason), metav1.PatchOptions{})
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+}
+
+func waitForApiToShutdown(f *framework.Framework) {
+	// config := f.ClientConfig()
+	config, err := framework.LoadConfig()
+	framework.ExpectNoError(err)
+
+	for i := 0; i < 10; i++ {
+		ctx, cancel := context.WithCancel(context.Background())
+		m := monitor.NewMonitorWithInterval(time.Second)
+		err = monitor.StartKubeAPIMonitoringWithNewConnections(ctx, m, config, 1*time.Second)
+		framework.ExpectNoError(err, "unable to monitor API")
+
+		// start := time.Now().UTC()
+		m.StartSampling(ctx)
+
+		time.Sleep(90 * time.Second)
+		cancel()
+		// end := time.Now().UTC()
+
+		conds := m.Conditions(time.Time{}, time.Time{})
+		framework.Logf("Conditions: %v", conds)
+		events := m.Intervals(time.Time{}, time.Time{})
+		framework.Logf("Events: %v", events)
+		disruption.FrameworkEventIntervals(f, events)
+		framework.Logf("Events: %v", events)
+		duration := events.Duration(0)
+		framework.Logf("Duration: %v", duration)
+		if float64(duration) == 0.0 {
+			return
+		}
+
+		framework.Logf("API is still not down on the %d iteration", i)
+	}
+
+	framework.Failf("kube-apiserver didn't shutdown")
+}
+
+var _ = ginkgo.Describe("[sig-api-machinery] Cluster resiliency", func() {
+	f := framework.NewDefaultFramework("single-node")
+	f.SkipNamespaceCreation = true
+	f.SkipPrivilegedPSPBinding = true
+
+	oc := exutil.NewCLIWithoutNamespace("cluster-resiliency")
+
+	ginkgo.It("Verify that OpenShift components are resilient to kubeapi server rollback", func() {
+		controlPlaneTopology, infraTopology := getTopologies(f)
+
+		if controlPlaneTopology != v1.SingleReplicaTopologyMode && infraTopology != v1.SingleReplicaTopologyMode {
+			e2eskipper.Skipf("Test is only relevant for single replica topologies")
+		}
+
+		disruption.Run(f, "Cluster Resiliency to API shutdown", "cluster-resiliency", disruption.TestData{},
+			[]upgrades.Test{
+				&controlplane.ResiliencyTest{},
```

Comment:
I guess you'll want to have one test for checking that API was down for less than 60 seconds, and a separate test to make sure components didn't restart during that time

Reply:
yes, we'll have several tests